typing/ikinds/ldd_hybrid_opt.ml

```diff
@@ -0,0 +1,600 @@
+      if lo' == nb.lo && hi' == nb.hi then other else node nb.v lo' hi'
+
+  (* --------- public constructors --------- *)
+  let[@inline] const (c : C.t) = leaf c
+
+  let[@inline] mk_var (v : var) : node = v.var_node
+
+  let rigid (name : V.t) = Var.make_rigid ~name ()
+
+  let new_var () = Var.make_var ()
+
+  (* --------- restrictions (x ← ⊥ / ⊤) (no memoization) --------- *)
+  let rec restrict0 (x : var) (w : node) : node =
+    if is_leaf w
+    then w
+    else
+      let n = node_block w in
+      let cmp = compare_var x n.v in
+      if cmp < 0
+      then w
+      else if cmp = 0
+      then n.lo
+      else
+        let lo' = restrict0 x n.lo in
+        let hi' = restrict0 x n.hi in
+        if lo' == n.lo && hi' == n.hi then w else node n.v lo' hi'
+
+  let rec restrict1 (x : var) (w : node) : node =
+    if is_leaf w
+    then w
+    else
+      let n = node_block w in
+      let cmp = compare_var x n.v in
+      if cmp < 0
+      then w
+      else if cmp = 0
+      then join n.lo n.hi
+      else
+        let lo' = restrict1 x n.lo in
+        let hi' = restrict1 x n.hi in
+        if lo' == n.lo && hi' == n.hi then w else node n.v lo' hi'
+
+  (* --------- force (no memoization) --------- *)
+  let rec force (w : node) : node =
+    if is_leaf w
+    then w
+    else
+      let n = node_block w in
+      if n.v.id > Var.rigid_var_start
+      then w
+      else
+        let lo' = force n.lo in
+        let hi' = force n.hi in
+        match n.v.state with
+        | Solved d ->
+          let d' = force d in
+          n.v.state <- Solved d';
+          join lo' (meet hi' d')
+        | Unsolved ->
+          if lo' == n.lo && hi' == n.hi
+          then w
+          else
+            let d' = mk_var n.v in
+            join lo' (meet hi' d')
+        | Rigid _ -> failwith "force: rigid variable shouldn't be here"
+
+  and var (v : var) = mk_var v
+
+  (* This function is equivalent to `restrict0 x (force w)` *)
+  let rec restrict0_force (x : var) (w : node) : node =
+    if x.id > Var.rigid_var_start
+    then restrict0 x (force w)
+    else if is_leaf w
+    then w
+    else
+      let n = node_block w in
+      match n.v.state with
+      | Solved d ->
+        let lo' = restrict0_force x n.lo in
+        let hi' = restrict0_force x n.hi in
+        let d_forced = force d in
+        n.v.state <- Solved d_forced;
+        let d' = restrict0 x d_forced in
+        join lo' (meet hi' d')
+      | Unsolved ->
+        let lo' = restrict0_force x n.lo in
+        if compare_var n.v x = 0
+        then lo'
+        else
+          let hi' = restrict0_force x n.hi in
+          if lo' == n.lo && hi' == n.hi
+          then w
+          else
+            let d' = mk_var n.v in
+            join lo' (meet hi' d')
+      | Rigid _ -> w
+
+  let sub_subsets (a : node) (b : node) = canonicalize (force a) (force b)
+
+  let sub_subsets_forced (a : node) (b : node) : node = canonicalize a b
```

Comment:
I think all usages of `sub_subsets_forced` can just be replaced by `sub_subsets`